plugins/destination/clickhouse/client/migrate.go

```diff
@@ -12,67 +14,140 @@ import (
+	var result []string
+	var tableChanges [][]schema.TableColumnChange
+	for _, t := range tables {
+		current := currentTables.Get(t.Name)
+		if current == nil {
+			continue
+		}
+		changes := t.GetChanges(current)
+		if !c.canSafeMigrate(changes) {
+			result = append(result, t.Name)
+			tableChanges = append(tableChanges, changes)
+		}
+	}
+	return result, tableChanges
+}
+
+func (*Client) canSafeMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && change.Current.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeRemove && change.Previous.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeUpdate {
+			return false
+		}
+	}
+	return true
+}
+
 func (c *Client) createTable(ctx context.Context, table *schema.Table) (err error) {
 	c.logger.Debug().Str("table", table.Name).Msg("Table doesn't exist, creating")
 
 	return c.conn.Exec(ctx, queries.CreateTable(table))
 }
 
-func (c *Client) autoMigrateTable(ctx context.Context, current, table *schema.Table) (err error) {
-	tableName := table.Name
-	c.logger.Info().Str("table", tableName).Msg("Table exists, migrating")
+func (c *Client) dropTable(ctx context.Context, table *schema.Table) (err error) {
+	c.logger.Debug().Str("table", table.Name).Msg("Dropping table")
+
+	return c.conn.Exec(ctx, queries.DropTable(table))
+}
 
-	have, want := current.GetChangedColumns(table)
-	if len(have) > 0 && c.spec.MigrateMode != specs.MigrateModeForced {
-		return fmt.Errorf("table %s has different types for columns %v but schema wants %v , add `migrate_mode: forced` to the destination spec to drop the columns", table.Name, have, want)
+func needsTableDrop(change schema.TableColumnChange) bool {
+	// We can safely add a nullable column without dropping the table
+	if change.Type == schema.TableColumnChangeTypeAdd && !change.Current.CreationOptions.NotNull {
+		return false
 	}
 
-	for _, column := range want {
-		if err := c.conn.Exec(ctx, queries.DropColumn(tableName, &column)); err != nil {
-			return err
-		}
-		if err := c.conn.Exec(ctx, queries.AddColumn(tableName, &column)); err != nil {
-			return err
-		}
+	// We can safely drop a nullable and add a nullable column without dropping the table
+	if change.Type == schema.TableColumnChangeTypeUpdate && !change.Current.CreationOptions.NotNull {
+		return false
 	}
 
-	for _, column := range table.GetAddedColumns(current) {
-		if err := c.conn.Exec(ctx, queries.AddColumn(tableName, &column)); err != nil {
-			return err
+	if change.Type == schema.TableColumnChangeTypeRemove {
+		return false
+	}
+
+	return true
+}
+
+func changesSorter(a, b schema.TableColumnChange) bool {
+	aDropTable := needsTableDrop(a)
+	bDropTable := needsTableDrop(b)
+	return aDropTable && !bDropTable
```

Comment:
may skip internal variables, just use call results

Reply:
I actually find this easier to debug the code this way, but don't have a strong opinion

Reply:
Done in https://github.com/cloudquery/cloudquery/pull/8255/commits/deb21b84f1b09ffa2ff78105209d3790e00bb6a3